ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Block/SupportsPeras.hs

```diff
@@ -108,13 +192,92 @@ class
 
   data PerasCert blk
 
+  data PerasVote blk
+
   data PerasValidationErr blk
 
   validatePerasCert ::
     PerasCfg blk ->
     PerasCert blk ->
     Either (PerasValidationErr blk) (ValidatedPerasCert blk)
 
+  validatePerasVote ::
+    PerasCfg blk ->
+    PerasVote blk ->
+    PerasVoteStakeDistr ->
+    Either (PerasValidationErr blk) (ValidatedPerasVote blk)
+
+const_PERAS_QUORUM_THRESHOLD :: PerasVoteStake
+const_PERAS_QUORUM_THRESHOLD = PerasVoteStake 0.75
+
+data PerasVoteAggregate blk = PerasVoteAggregate
+  { pvaTarget :: !(PerasVoteTarget blk)
+  , pvaVotes :: !(Set (WithArrivalTime (ValidatedPerasVote blk)))
+  , pvaTotalStake :: !PerasVoteStake
+  , pvaMaybeCert :: !(Maybe (ValidatedPerasCert blk))
+  }
+  deriving stock (Generic, Eq, Ord, Show)
+  deriving anyclass NoThunks
+
+emptyPerasVoteAggregate :: PerasVoteTarget blk -> PerasVoteAggregate blk
+emptyPerasVoteAggregate target =
+  PerasVoteAggregate
+    { pvaTotalStake = PerasVoteStake 0
+    , pvaTarget = target
+    , pvaVotes = Set.empty
+    , pvaMaybeCert = Nothing
+    }
+
+data UpdatePerasVoteAggregateResult blk
+  = IncorrectPerasVoteTarget
+  | AddedPerasVoteButDidntGenerateNewCert (PerasVoteAggregate blk)
+  | AddedPerasVoteAndGeneratedNewCert (PerasVoteAggregate blk) (ValidatedPerasCert blk)
+  deriving stock (Generic, Eq, Ord, Show)
+  deriving anyclass NoThunks
+
+updatePerasVoteAggregate ::
+  StandardHash blk =>
+  PerasVoteAggregate blk ->
+  WithArrivalTime (ValidatedPerasVote blk) ->
+  UpdatePerasVoteAggregateResult blk
+updatePerasVoteAggregate
+  pva@PerasVoteAggregate
+    { pvaTarget = (roundNo, point)
+    , pvaVotes = existingVotes
+    , pvaTotalStake = initialStake
+    , pvaMaybeCert = mExistingCert
+    }
+  vote =
+    if getPerasVoteRound vote == roundNo && getPerasVoteVotedBlock vote == point
+      then
+        let pvaTotalStake = initialStake + vpvVoteStake (forgetArrivalTime vote)
+            pvaVotes = Set.insert vote existingVotes
```

Comment:
This insert not actually inserting anything (because the vote already exists) needs to be taken into consideration when updating the stake.